src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,310 @@
+// SPDX-FileCopyrightText: © 2022 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.6.12;
+
+import {GemAbstract} from "dss-interfaces/ERC/GemAbstract.sol";
+import {DaiAbstract} from "dss-interfaces/dss/DaiAbstract.sol";
+import {PsmAbstract} from "dss-interfaces/dss/PsmAbstract.sol";
+import {GemJoinAbstract} from "dss-interfaces/dss/GemJoinAbstract.sol";
+
+/**
+ * @author Lev Livnev <lev@liv.nev.org.uk>
+ * @author Nazar Duchak <nazar@clio.finance>
+ * @title An Input Conduit for real-world assets (RWA).
+ * @dev This contract differs from the original [RwaInputConduit](https://github.com/makerdao/MIP21-RWA-Example/blob/fce06885ff89d10bf630710d4f6089c5bba94b4d/src/RwaConduit.sol#L20-L39):
+ *  - The caller of `push()` is not required to hold MakerDAO governance tokens.
+ *  - The `push()` method is permissioned.
+ *  - `push()` permissions are managed by `mate()`/`hate()` methods.
+ *  - Require PSM address in constructor
+ *  - The `push()` method swaps GEM to DAI using PSM
+ *  - THe `push()` method with `amt` argument swaps specified amount of GEM to DAI using PSM
+ *  - The `quit` method allows moving outstanding GEM balance to `quitTo`. It can be called only by the admin.
+ *  - The `quit` method with `amount` argument allows moving specified amount of GEM balance to `quitTo`.
+ *  - The `file` method allows updating `quitTo`, `to` addresses. It can be called only by the admin.
+ */
+contract RwaInputConduit3 {
+    /// @notice PSM GEM token contract address
+    GemAbstract public immutable gem;
+    /// @notice PSM contract address
+    PsmAbstract public immutable psm;
+    /// @dev DAI/GEM resolution difference.
+    uint256 private immutable to18ConvertionFactor;
+
+    /// @dev This is declared here so the storage layout lines up with RwaInputConduit.
+    address private __unused_gov;
+    /// @notice Dai token contract address
+    DaiAbstract public dai;
+    /// @notice RWA urn contract address
+    address public to;
+
+    /// @notice Addresses with admin access on this contract. `wards[usr]`
+    mapping(address => uint256) public wards;
+    /// @notice Addresses with push access on this contract. `may[usr]`
+    mapping(address => uint256) public may;
+
+    /// @notice Exit address
```

Comment:
Also here, please expand the explanation as done for the output conduit.

Reply:
Done